File: sdk/react-native/src/ValydarClient.ts
import type {
  ValydarConfig,
  VerificationResponse,
  DocumentUploadResponse,
  FaceMatchResponse,
  LivenessResult,
} from './types';

export class ValydarClient {
  private apiKey: string;
  private baseUrl: string;

  constructor(config: ValydarConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.apiUrl ?? 'https://api.dev.valydar.com';
  }

  private async request<T>(
    method: string,
    path: string,
    body?: object | FormData,
  ): Promise<T> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.apiKey}`,
    };

    const isFormData = body instanceof FormData;
    if (!isFormData) {
      headers['Content-Type'] = 'application/json';
    }

    const res = await fetch(`${this.baseUrl}${path}`, {
      method,
      headers,
      body: body ? (isFormData ? body : JSON.stringify(body)) : undefined,
    });

    if (!res.ok) {
      const err = await res.json().catch(() => ({ error: res.statusText }));
      throw new Error(err.error?.message ?? `HTTP ${res.status}`);
    }

    return res.json();
  }

  async createVerification(
    clientReference?: string,
    checks?: string[],
  ): Promise<VerificationResponse> {
    return this.request<VerificationResponse>('POST', '/verifications', {
      client_reference: clientReference ?? `rn_${Date.now()}`,
      checks: checks ?? ['document', 'face_match', 'liveness'],
    });
  }

  async getVerification(id: string): Promise<VerificationResponse> {
    return this.request<VerificationResponse>('GET', `/verifications/${id}`);
  }

  async uploadDocument(
    verificationId: string,
    imageUri: string,
    documentType?: string,
  ): Promise<DocumentUploadResponse> {
    const form = new FormData();
    form.append('file', {
      uri: imageUri,
      type: 'image/jpeg',
      name: 'document.jpg',
    } as unknown as Blob);
    if (documentType) {
      form.append('document_type', documentType);
    }
    return this.request<DocumentUploadResponse>(
      'POST',
      `/verifications/${verificationId}/documents`,
      form,
    );
  }

  async uploadSelfie(
    verificationId: string,
    imageUri: string,
  ): Promise<{ document_id: string }> {
    const form = new FormData();
    form.append('file', {
      uri: imageUri,
      type: 'image/jpeg',
      name: 'selfie.jpg',
    } as unknown as Blob);
    return this.request<{ document_id: string }>(
      'POST',
      `/verifications/${verificationId}/selfie`,
      form,
    );
  }

  async faceMatch(
    verificationId: string,
    documentId: string,
    selfieId: string,
  ): Promise<FaceMatchResponse> {
    return this.request<FaceMatchResponse>(
      'POST',
      `/verifications/${verificationId}/face-match`,
      { document_id: documentId, selfie_id: selfieId },
    );
  }

  async documentLiveness(
    verificationId: string,
    documentId: string,
  ): Promise<LivenessResult> {
    return this.request<LivenessResult>(
      'POST',
      `/verifications/${verificationId}/documents/${documentId}/liveness`,
    );
  }

  async health(): Promise<{ status: string }> {
    return this.request<{ status: string }>('GET', '/health');
  }
}

File: sdk/flutter/lib/valydar_client.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

class ValydarClient {
  final String apiKey;
  final String baseUrl;

  ValydarClient({required this.apiKey, this.baseUrl = 'https://api.dev.valydar.com'});

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    http.MultipartRequest? multipart,
  }) async {
    final uri = Uri.parse('$baseUrl$path');
    final headers = <String, String>{
      'Authorization': 'Bearer $apiKey',
    };

    http.Response response;

    if (multipart != null) {
      multipart.headers.addAll(headers);
      final streamed = await multipart.send();
      response = await http.Response.fromStream(streamed);
    } else if (body != null) {
      headers['Content-Type'] = 'application/json';
      response = await http.post(uri, headers: headers, body: jsonEncode(body));
    } else {
      response = await http.get(uri, headers: headers);
    }

    if (response.statusCode >= 400) {
      final err = jsonDecode(response.body);
      throw Exception(err['error']?['message'] ?? 'HTTP ${response.statusCode}');
    }

    return jsonDecode(response.body);
  }

  Future<Map<String, dynamic>> createVerification({
    String? clientReference,
    List<String>? checks,
  }) {
    return _request('POST', '/verifications', body: {
      'client_reference': clientReference ?? 'flutter_${DateTime.now().millisecondsSinceEpoch}',
      'checks': checks ?? ['document', 'face_match', 'liveness'],
    });
  }

  Future<Map<String, dynamic>> getVerification(String id) {
    return _request('GET', '/verifications/$id');
  }

  Future<Map<String, dynamic>> uploadDocument(
    String verificationId,
    File image,
    String? documentType,
  ) async {
    final uri = Uri.parse('$baseUrl/verifications/$verificationId/documents');
    final request = http.MultipartRequest('POST', uri);
    request.headers['Authorization'] = 'Bearer $apiKey';
    request.files.add(await http.MultipartFile.fromPath('file', image.path));
    if (documentType != null) {
      request.fields['document_type'] = documentType;
    }
    final streamed = await request.send();
    final response = await http.Response.fromStream(streamed);
    if (response.statusCode >= 400) {
      final err = jsonDecode(response.body);
      throw Exception(err['error']?['message'] ?? 'HTTP ${response.statusCode}');
    }
    return jsonDecode(response.body);
  }

  Future<Map<String, dynamic>> uploadSelfie(
    String verificationId,
    File image,
  ) async {
    final uri = Uri.parse('$baseUrl/verifications/$verificationId/selfie');
    final request = http.MultipartRequest('POST', uri);
    request.headers['Authorization'] = 'Bearer $apiKey';
    request.files.add(await http.MultipartFile.fromPath('file', image.path));
    final streamed = await request.send();
    final response = await http.Response.fromStream(streamed);
    if (response.statusCode >= 400) {
      final err = jsonDecode(response.body);
      throw Exception(err['error']?['message'] ?? 'HTTP ${response.statusCode}');
    }
    return jsonDecode(response.body);
  }

  Future<Map<String, dynamic>> faceMatch(
    String verificationId,
    String documentId,
    String selfieId,
  ) {
    return _request('POST', '/verifications/$verificationId/face-match', body: {
      'document_id': documentId,
      'selfie_id': selfieId,
    });
  }

  Future<Map<String, dynamic>> documentLiveness(
    String verificationId,
    String documentId,
  ) {
    return _request(
      'POST',
      '/verifications/$verificationId/documents/$documentId/liveness',
    );
  }

  Future<Map<String, dynamic>> health() {
    return _request('GET', '/health');
  }
}
